app/clients/letter/dvla.py

```diff
@@ -80,5 +103,99 @@ def authenticate(self):
+        """
+        from app.models import INTERNATIONAL_POSTAGE_TYPES
+
+        if postage in INTERNATIONAL_POSTAGE_TYPES:
+            raise NotImplementedError
+
+        current_app.logger.info(f"Sending letter with id {notification_id}")
+
+        try:
+            response = requests.post(
+                f"{current_app.config['DVLA_API_BASE_URL']}/print-request/v1/print/jobs",
+                headers=self._get_auth_headers(),
+                json=self.format_create_print_job_json(
+                    notification_id, address, postage, service_id, organisation_id, pdf_file
+                ),
+            )
+            response.raise_for_status()
+        except requests.HTTPError as e:
+            # Catch errors and raise our own to indicate what action to take.
+            # If the error has details, we add them to the error message.
+            if e.response.status_code == 400:
+                raise DvlaException(response.json()["errors"][0]["detail"])
+            elif e.response.status_code in {401, 403}:
+                raise DvlaUnauthorisedRequestException(
+                    f'{e.response.status_code} {response.json()["errors"][0]["detail"]}'
+                )
+            elif e.response.status_code == 409:
+                raise DvlaDuplicatePrintRequestException(response.json()["errors"][0]["detail"])
+            elif e.response.status_code == 429:
+                raise DvlaThrottlingException(str(e))
+            elif e.response.status_code >= 500:
+                raise DvlaRetryableExcepton(str(e))
+            else:
+                raise DvlaException(str(e))
+        else:
+            return response.json()
+
+    def format_create_print_job_json(
+        self, notification_id, address_lines, postage, service_id, organisation_id, pdf_file
+    ):
+        from app.models import FIRST_CLASS
+
+        recipient_name = address_lines.pop(0)
+
+        json_payload = {
+            "id": notification_id,
+            "standardParams": {
+                "jobType": "NOTIFY",
+                "templateReference": "NOTIFY",
+                "businessIdentifier": notification_id,
+                "recipientName": recipient_name,
+                "address": {"unstructuredAddress": self.build_unstructured_address(address_lines)},
+            },
+            "customParams": [
+                {"key": "pdfContent", "value": base64.b64encode(pdf_file).decode("utf-8")},
+                {"key": "organisationIdentifier", "value": organisation_id},
+                {"key": "serviceIdentifier", "value": service_id},
+            ],
+        }
+
+        if postage == FIRST_CLASS:
+            json_payload["standardParams"]["despatchMethod"] = "FIRST"
+
+        return json_payload
+
+    def build_unstructured_address(self, address_lines_without_recipient):
+        address_line_keys = ["line1", "line2", "line3", "line4", "line5"]
+        postcode = address_lines_without_recipient.pop()
```

Comment:
This is mutating `address_lines_without_recipient` which was passed in - if anything in the caller tries to use it after calling this function it will (probably unexpectedly) not have the postcode in any more.

Might be nice to make a copy of this dict before modifying it so that the input isn't altered (ie it's a pure function).

We can also mark this as a `staticmethod` as it doesn't need `self`.

Reply:
I've updated the places that were using `pop` to get values by using list slicing instead